File: packages/sdk/src/providers/monnify.ts
import type {
  Provider,
  TransactionConfig,
  TransactionResult,
  VerificationResult,
  ChargeConfig,
  ChargeResult,
  RefundConfig,
  RefundResult,
  WebhookEvent,
  PayVaultConfig,
} from '../types';
import { HttpClient } from '../http';
import {
  generateReference,
  normalizeStatus,
  normalizeChannel,
  hmacSha512,
} from '../utils';
import { ValidationError, TransactionError, PayVaultError, ProviderError } from '../errors';

const MONNIFY_SANDBOX_URL = 'https://sandbox.monnify.com/api';
const MONNIFY_LIVE_URL = 'https://api.monnify.com/api';

/**
 * Parse pipe-delimited Monnify credentials.
 * Format: `apiKey|secretKey|contractCode`
 */
function parseMonnifyCredentials(secret: string): {
  apiKey: string;
  secretKey: string;
  contractCode: string;
} {
  const parts = secret.split('|');
  if (parts.length !== 3) {
    throw new PayVaultError(
      'Monnify credentials must be in the format: apiKey|secretKey|contractCode',
      { code: 'VALIDATION_ERROR', provider: 'monnify' }
    );
  }
  const [apiKey, secretKey, contractCode] = parts.map(p => p.trim());
  if (!apiKey || !secretKey || !contractCode) {
    throw new PayVaultError(
      'Monnify credentials must include apiKey, secretKey, and contractCode',
      { code: 'VALIDATION_ERROR', provider: 'monnify' }
    );
  }
  return { apiKey, secretKey, contractCode };
}

/**
 * Map Monnify payment status to unified status.
 */
function mapMonnifyStatus(status: string): 'success' | 'failed' | 'pending' | 'abandoned' {
  switch (status?.toUpperCase()) {
    case 'PAID':
    case 'OVERPAID':
      return 'success';
    case 'FAILED':
    case 'EXPIRED':
    case 'CANCELLED':
      return 'failed';
    default:
      return 'pending';
  }
}

export class MonnifyProvider implements Provider {
  name = 'monnify' as const;
  private http: HttpClient;
  private rawSecret: string;
  private webhookSecret?: string;
  private baseUrl: string;
  private defaultCurrency: string;
  private defaultMetadata: Record<string, any>;

  constructor(config: PayVaultConfig) {
    this.rawSecret = config.secretKey;
    this.webhookSecret = config.webhookSecret;
    this.defaultCurrency = config.currency || 'NGN';
    this.defaultMetadata = config.metadata || {};

    // Determine base URL: test keys start with MK_TEST_
    const { apiKey } = parseMonnifyCredentials(config.secretKey);
    this.baseUrl = config.baseUrl || (apiKey.startsWith('MK_TEST_') ? MONNIFY_SANDBOX_URL : MONNIFY_LIVE_URL);

    this.http = new HttpClient('monnify', {
      retry: config.retry,
      timeout: config.timeout,
    });
  }

  /**
   * Fetch an access token using Basic Auth (apiKey:secretKey → base64).
   * Monnify requires this on every request cycle.
   */
  private async fetchAccessToken(): Promise<string> {
    const { apiKey, secretKey } = parseMonnifyCredentials(this.rawSecret);
    const credentials = Buffer.from(`${apiKey}:${secretKey}`).toString('base64');

    const response = await this.http.post(
      `${this.baseUrl}/v1/auth/login`,
      {},
      { 'Authorization': `Basic ${credentials}` }
    );

    const token: string = response.data?.responseBody?.accessToken;
    if (!token) {
      throw new ProviderError('Failed to obtain Monnify access token', 'monnify', 401, response.data);
    }
    return token;
  }

  private async authHeaders(): Promise<Record<string, string>> {
    const token = await this.fetchAccessToken();
    return { 'Authorization': `Bearer ${token}` };
  }

  async initializeTransaction(config: TransactionConfig): Promise<TransactionResult> {
    if (!config.email) {
      throw new ValidationError('Email is required', 'monnify', 'email');
    }
    if (!config.amount || config.amount <= 0) {
      throw new ValidationError('Amount must be greater than 0', 'monnify', 'amount');
    }

    const { contractCode } = parseMonnifyCredentials(this.rawSecret);
    const currency = config.currency || this.defaultCurrency;
    const reference = config.reference || generateReference('pvt_mn');

    const payload: Record<string, any> = {
      amount: config.amount, // Monnify uses major units (naira, not kobo)
      customerName: config.customer
        ? `${config.customer.firstName || ''} ${config.customer.lastName || ''}`.trim() || config.email
        : config.email,
      customerEmail: config.email,
      paymentReference: reference,
      paymentDescription: 'Payment',
      currencyCode: currency,
      contractCode,
      redirectUrl: config.callbackUrl || '',
      paymentMethods: ['CARD', 'ACCOUNT_TRANSFER'],
      metaData: { ...this.defaultMetadata, ...config.metadata },
    };

    if (config.channels) {
      payload.paymentMethods = config.channels.map(ch => this.mapChannel(ch));
    }

    const headers = await this.authHeaders();
    const response = await this.http.post(
      `${this.baseUrl}/v1/merchant/transactions/init-transaction`,
      payload,
      headers
    );

    const resData = response.data?.responseBody;

    return {
      success: response.data?.requestSuccessful === true,
      provider: 'monnify',
      authorizationUrl: resData?.checkoutUrl || '',
      accessCode: reference,
      reference,
      raw: response.data,
    };
  }

  async verifyTransaction(reference: string): Promise<VerificationResult> {
    if (!reference) {
      throw new ValidationError('Reference is required', 'monnify', 'reference');
    }

    const headers = await this.authHeaders();
    // Monnify expects URL-encoded transactionReference
    const response = await this.http.get(
      `${this.baseUrl}/v2/transactions/${encodeURIComponent(reference)}`,
      headers
    );
    const data = response.data?.responseBody;
    const currency = data?.currencyCode || this.defaultCurrency;

    return {
      success: mapMonnifyStatus(data?.paymentStatus) === 'success',
      status: mapMonnifyStatus(data?.paymentStatus),
      provider: 'monnify',
      reference: data?.paymentReference || reference,
      amount: data?.amountPaid ?? data?.amount ?? 0,
      currency,
      channel: normalizeChannel(data?.paymentMethod?.toLowerCase() || 'card') as any,
      paidAt: data?.paidOn || null,
      customer: {
        email: data?.customer?.email || data?.customerEmail || '',
        firstName: data?.customer?.name?.split(' ')[0],
        lastName: data?.customer?.name?.split(' ').slice(1).join(' '),
      },
      fees: data?.payableAmount ? data.payableAmount - (data.amountPaid ?? 0) : undefined,
      raw: response.data,
    };
  }

  async charge(config: ChargeConfig): Promise<ChargeResult> {
    // Monnify doesn't support direct charges (no card tokenization via API)
    // All payments go through the redirect/checkout flow
    throw new PayVaultError(
      'Monnify does not support direct charges. Use initializeTransaction() for the checkout redirect flow.',
      { code: 'UNSUPPORTED_OPERATION', provider: 'monnify' }
    );
  }

  async submitAuthorization(
    _reference: string,
    _auth: { type: string; value: string }
  ): Promise<ChargeResult> {
    throw new PayVaultError(
      'Monnify does not support OTP/PIN authorization. All auth is handled via checkout redirect.',
      { code: 'UNSUPPORTED_OPERATION', provider: 'monnify' }
    );
  }

  async refund(_config: RefundConfig): Promise<RefundResult> {
    throw new PayVaultError(
      'Monnify refunds must be processed manually via the Monnify merchant dashboard.',
      { code: 'UNSUPPORTED_OPERATION', provider: 'monnify' }
    );
  }

  verifyWebhook(payload: string | Buffer, signature: string): boolean {
    const { secretKey } = parseMonnifyCredentials(this.rawSecret);
    const secret = this.webhookSecret || secretKey;
    const hash = hmacSha512(
      typeof payload === 'string' ? payload : payload.toString('utf8'),
      secret
    );
    return hash === signature.toLowerCase();
  }

  parseWebhook(payload: string | Buffer): WebhookEvent {
    const body =
      typeof payload === 'string'
        ? JSON.parse(payload)
        : JSON.parse(payload.toString('utf8'));

    const data = body.eventData || body.responseBody || {};
    const currency = data.currency || data.currencyCode || this.defaultCurrency;

    return {
      id: data.transactionReference || '',
      provider: 'monnify',
      type: body.eventType || 'SUCCESSFUL_TRANSACTION',
      reference: data.paymentReference || data.transactionReference || '',
      status: mapMonnifyStatus(data.paymentStatus),
      amount: data.amountPaid ?? data.settlementAmount ?? 0,
      currency,
      customer: {
        email: data.customer?.email || data.customerEmail || '',
        firstName: data.customer?.name?.split(' ')[0],
        lastName: data.customer?.name?.split(' ').slice(1).join(' '),
      },
      timestamp: data.paidOn || new Date().toISOString(),
      raw: body,
    };
  }

  private mapChannel(channel: string): string {
    const map: Record<string, string> = {
      card: 'CARD',
      bank_transfer: 'ACCOUNT_TRANSFER',
      ussd: 'USSD',
      mobile_money: 'PHONE_NUMBER',
    };
    return map[channel] || channel.toUpperCase();
  }
}

File: apps/api/internal/services/monnify.go
package services

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha512"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"strings"
	"time"
)

type MonnifyProvider struct {
	client *http.Client
}

func NewMonnifyProvider() *MonnifyProvider {
	return &MonnifyProvider{
		client: &http.Client{Timeout: 30 * time.Second},
	}
}

func (m *MonnifyProvider) Name() string { return "monnify" }

func (m *MonnifyProvider) InitiateCharge(ctx context.Context, req ChargeRequest) (*ChargeResponse, error) {
	rawSecret, _ := ctx.Value(ctxKeySecretKey).(string)
	apiKey, secretKey, contractCode, err := parseMonnifyCredentials(rawSecret)
	if err != nil {
		return nil, fmt.Errorf("monnify: %w", err)
	}

	accessToken, err := m.fetchAccessToken(ctx, apiKey, secretKey)
	if err != nil {
		return nil, fmt.Errorf("monnify initiate charge auth: %w", err)
	}

	body := map[string]interface{}{
		"amount":             float64(req.AmountKobo) / 100.0,
		"customerName":       req.Email,
		"customerEmail":      req.Email,
		"paymentReference":   req.Reference,
		"paymentDescription": "Payment",
		"currencyCode":       req.Currency,
		"contractCode":       contractCode,
		"redirectUrl":        req.CallbackURL,
		"paymentMethods":     []string{"CARD", "ACCOUNT_TRANSFER"},
	}

	data, err := m.doRequest(ctx, "POST", "/v1/merchant/transactions/init-transaction", body, accessToken, apiKey)
	if err != nil {
		return nil, fmt.Errorf("monnify initiate charge: %w", err)
	}

	resData, _ := data["responseBody"].(map[string]interface{})
	return &ChargeResponse{
		ProviderRef: getString(resData, "transactionReference"),
		AuthURL:     getString(resData, "checkoutUrl"),
		Status:      "pending",
		RawResponse: data,
	}, nil
}

func (m *MonnifyProvider) VerifyTransaction(ctx context.Context, providerRef string) (*VerifyResponse, error) {
	rawSecret, _ := ctx.Value(ctxKeySecretKey).(string)
	apiKey, secretKey, _, err := parseMonnifyCredentials(rawSecret)
	if err != nil {
		return nil, fmt.Errorf("monnify: %w", err)
	}

	accessToken, err := m.fetchAccessToken(ctx, apiKey, secretKey)
	if err != nil {
		return nil, fmt.Errorf("monnify verify auth: %w", err)
	}

	data, err := m.doRequest(ctx, "GET", "/v2/transactions/"+providerRef, nil, accessToken, apiKey)
	if err != nil {
		return nil, fmt.Errorf("monnify verify: %w", err)
	}

	resData, _ := data["responseBody"].(map[string]interface{})
	amountPaid, _ := resData["amountPaid"].(float64)

	return &VerifyResponse{
		ProviderRef: getString(resData, "transactionReference"),
		Status:      mapMonnifyStatus(getString(resData, "paymentStatus")),
		AmountKobo:  int64(math.Round(amountPaid * 100)),
		Currency:    getString(resData, "currencyCode"),
		PaidAt:      getString(resData, "paidOn"),
		Channel:     getString(resData, "paymentMethod"),
		RawResponse: data,
	}, nil
}

func (m *MonnifyProvider) Refund(ctx context.Context, providerRef string, amountKobo int64) (*RefundResponse, error) {
	return nil, fmt.Errorf("monnify: refunds must be processed manually via the Monnify merchant dashboard")
}

func (m *MonnifyProvider) VerifyWebhookSignature(payload []byte, signature string, secret string) bool {
	_, secretKey, _, err := parseMonnifyCredentials(secret)
	if err != nil {
		return false
	}

	mac := hmac.New(sha512.New, []byte(secretKey))
	mac.Write(payload)
	expected := hex.EncodeToString(mac.Sum(nil))
	return hmac.Equal([]byte(expected), []byte(strings.ToLower(signature)))
}

func (m *MonnifyProvider) fetchAccessToken(ctx context.Context, apiKey, secretKey string) (string, error) {
	req, err := http.NewRequestWithContext(ctx, "POST", monnifyBaseURL(apiKey)+"/v1/auth/login", nil)
	if err != nil {
		return "", err
	}

	credentials := base64.StdEncoding.EncodeToString([]byte(apiKey + ":" + secretKey))
	req.Header.Set("Authorization", "Basic "+credentials)

	resp, err := m.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return "", fmt.Errorf("monnify auth: invalid JSON response: %s", string(respBody))
	}

	if success, ok := result["requestSuccessful"].(bool); ok && !success {
		msg := getString(result, "responseMessage")
		if msg == "" {
			msg = "authentication failed"
		}
		return "", fmt.Errorf("monnify auth API error: %s", msg)
	}

	resData, _ := result["responseBody"].(map[string]interface{})
	token := getString(resData, "accessToken")
	if token == "" {
		return "", fmt.Errorf("monnify auth: access token missing from response")
	}

	return token, nil
}

func (m *MonnifyProvider) doRequest(ctx context.Context, method, path string, body interface{}, accessToken, apiKey string) (map[string]interface{}, error) {
	var reqBody io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reqBody = bytes.NewReader(jsonBody)
	}

	req, err := http.NewRequestWithContext(ctx, method, monnifyBaseURL(apiKey)+path, reqBody)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := m.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("monnify: invalid JSON response: %s", string(respBody))
	}

	if success, ok := result["requestSuccessful"].(bool); ok && !success {
		msg := getString(result, "responseMessage")
		if msg == "" {
			msg = resp.Status
		}
		return nil, fmt.Errorf("monnify API error: %s", msg)
	}

	return result, nil
}

func monnifyBaseURL(apiKey string) string {
	if strings.HasPrefix(apiKey, "MK_TEST_") {
		return "https://sandbox.monnify.com/api"
	}
	return "https://api.monnify.com/api"
}

func parseMonnifyCredentials(secret string) (string, string, string, error) {
	parts := strings.SplitN(secret, "|", 3)
	if len(parts) != 3 {
		return "", "", "", fmt.Errorf("credentials must be encoded as apiKey|secretKey|contractCode")
	}
	apiKey := strings.TrimSpace(parts[0])
	secretKey := strings.TrimSpace(parts[1])
	contractCode := strings.TrimSpace(parts[2])
	if apiKey == "" || secretKey == "" || contractCode == "" {
		return "", "", "", fmt.Errorf("credentials must include apiKey, secretKey, and contractCode")
	}
	return apiKey, secretKey, contractCode, nil
}

func mapMonnifyStatus(status string) string {
	switch status {
	case "PAID", "OVERPAID":
		return "success"
	case "FAILED", "EXPIRED", "CANCELLED":
		return "failed"
	default:
		return "pending"
	}
}
